lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -517,43 +630,95 @@ def _cleanup_bootstrap_conf_if_applies(self) -> None:
         """Remove some conf props in the CM nodes that contributed to the cluster bootstrapping."""
         self.opensearch_config.cleanup_bootstrap_conf()
 
-    def on_allocation_exclusion_add_failed(self):
-        """Callback for when the OpenSearch service fails stopping."""
-        self.unit.status = BlockedStatus(AllocationExclusionFailed)
+    def _restart_unit_with_conf_if_needed(self):
+        """Trigger a restart event on the current unit if a recomputed conf was passed."""
+        node_conf_update_app = self.peers_data.get_object(Scope.APP, "update-node-config")
+        if not node_conf_update_app:
+            return
+
+        node = Node.from_dict(node_conf_update_app)
+        if node.name != self.unit_name:
+            return
 
-    def on_unassigned_shards(self, unassigned_shards: int):
-        """Called during node shutdown / horizontal scale-down if some shards left unassigned."""
-        self.app.status = MaintenanceStatus(HorizontalScaleUpSuggest.format(unassigned_shards))
+        prev_update_node = self.peers_data.get_object(Scope.UNIT, "last-update-node-config")
+        if prev_update_node:
+            prev_update_node = Node.from_dict(prev_update_node)
 
-    def append_allocation_exclusion_to_remove(self, unit_name) -> None:
-        """Store a unit in the relation data bag, to be removed from the allocation exclusion."""
-        if not self.unit.is_leader():
-            self.peers_data.put(Scope.UNIT, "remove_from_allocation_exclusions", unit_name)
+        # already updated
+        if node == prev_update_node:
             return
 
-        exclusions = set(
-            self.peers_data.get(Scope.APP, "remove_from_allocation_exclusions", "").split(",")
+        node_conf_update_unit = self.peers_data.get_object(Scope.UNIT, "update-node-config")
+        if not node_conf_update_unit:  # already emitted update / restart directive
+            self.peers_data.put_object(Scope.UNIT, "update-node-config", node_conf_update_app)
+
+        self.on[self.service_manager.name].acquire_lock.emit(
```

Comment:
This lock is removed once opensearch is fully restarted, right?

Reply:
Yes, once the `_restart_opensearch` method returns (successfully completes or gets deferred), the lock gets released. And then on a future execution (say the restart event got deferred for a timeout for ex) we test against the "update conf" of the node to see if we should request once again the lock or not